pkg/capacity/pci.go

```diff
@@ -0,0 +1,273 @@
+package capacity
+
+import (
+	"bufio"
+	"bytes"
+	_ "embed"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+const (
+	pciDir = "/sys/bus/pci/devices"
+)
+
+type Device struct {
+	ID   uint16
+	Name string
+}
+
+type Vendor struct {
+	ID      uint16
+	Name    string
+	Devices map[uint16]Device
+}
+
+var (
+	//go:embed pci/pci.ids
+	src []byte
+
+	pattern = regexp.MustCompile(`^(\t*)([\d|a-f]{4})(.+)$`)
+	data    map[uint16]Vendor
+
+	gpuVendorsWhitelist = []uint16{
+		0x1002, // AMD
+		0x10de, // NVIDIA
+	}
+)
+
+func init() {
+	var err error
+	data, err = loadVendors(src)
+	if err != nil {
+		panic(fmt.Sprintf("failed to build vendors db: %s", err))
+	}
+}
+
+func loadVendors(src []byte) (map[uint16]Vendor, error) {
+	vendors := make(map[uint16]Vendor)
+
+	scanner := bufio.NewScanner(bytes.NewBuffer(src))
+	var current *Vendor
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.HasPrefix(line, "#") {
+			continue
+		}
+
+		matches := pattern.FindStringSubmatch(line)
+		if matches == nil {
+			continue
+		}
+		// this should have `4 parts`
+		tabs := matches[1]
+		id64, err := strconv.ParseUint(matches[2], 16, 16)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse vendor or device id: %w", err)
+		}
+		rest := strings.TrimSpace(matches[3])
+
+		id := uint16(id64)
+		// the line has no leading tabs. this is a new vendor
+		if len(tabs) == 0 {
+			if current != nil {
+				vendors[current.ID] = *current
+			}
+			current = &Vendor{
+				ID:      id,
+				Name:    rest,
+				Devices: make(map[uint16]Device),
+			}
+		} else if len(tabs) == 1 {
+			// that is a device
+			if current == nil {
+				return nil, fmt.Errorf("device appeared before a vendor: %s", line)
+			}
+
+			current.Devices[id] = Device{
+				ID:   id,
+				Name: rest,
+			}
+		}
+	}
+
+	return vendors, nil
+}
+
+// GetVendor gets a vendor object given Vendor ID
+func GetVendor(vendor uint16) (Vendor, bool) {
+	v, ok := data[vendor]
```

Comment:
let's change to `return data[vendor]`

Reply:
Unfortunately it's not possible it's Go magic because in go both
```go
v, ok := data[vendor]
// and
v := data[vendor]
```
are valid, so when u try to do `return data[vendor]` it uses the `v := data[vendor]` symatic then fails to compile because the function returns 2 variables.